### DISCUSSION TITLE
using placeholders in autorization rules file

### DISCUSSION THREAD

**Discussion by rickou** [emqx/emqx#14391] [Q&A] [answered]
hi, 
i have a question, i enabled MQTT over TLS with peer verification. all is working regarding connection.
now i would like to define rules for publishing/subscribing depending on the client connected
I set an authorization file with rules in order to allow publishing only for logged-in client with username as a specific format (regex) and a topic path like "xxx/${cert_common_name}/yyy"
this also works.
but i also would like to check that the username and/or clientid = $cert_common_name... but  it seems i can't use the ${cert_common_name} placeholder in the "who" field... (it never match) or maybe i missed something... 
This works: {allow, {username, {re, "GBOX[0-9]{6}"}}, publish, ["gb/${cert_common_name}/data"]}.
This doesn't works: {allow, {username, "${cert_common_name}"}, publish, ["gb/${cert_common_name}/data"]}.

I would like a final rule like : {allow, {'and',[{username, "${cert_common_name}"}, {username, {re, "GBOX[0-9]{6}"}}]}, publish, ["gb/${cert_common_name}/data"]}.

**Comment by id** [comment]
Hi @rickou. You can verify that username and/or clientid match CN in the authentication phase, check out this example: https://github.com/emqx/emqx-usage-example/blob/main/client-attributes/config/authn.hocon

**Comment by zmstone** [comment, marked as answer]
The "who" match condition never supported templating.

If you want to allow A username to publish to topic `gb/${username}/data`,
you can simply write an ALL rule like this: `{allow, all, publish, ["gb/${username}/data"]}`. i.e. no need to use a template for `all`.

If you want to apply this only to clients who has username also matches regex pattern `GBOX[0-9]{6}`,
you will need to use client attributes.

Extract a client attribute like this:
```
export EMQX_MQTT__CLIENT_ATTRS_INIT="[{expression = \"iif(regex_match(coalesce(username,''),'GBOX[0-9]{6}'),username,'')\", set_as_attr = allowed_username}]"
```

And use a rule like this:
```
{allow, all, subscribe, ["gb/${client_attrs.allowed_username}/#"]}.
```

breakdown:

- `coalesce(username,'')`: if the client provided username, then use the username, otherwise empty string
- `regex_match(..., 'GBOX[0-9]{6}')`: if the username matches the pattern, then return 'true' otherwise 'false'
- `iif(Cond, TrueValue, FalseValue)`: return `TrueValue` if `Cond` is true, otherwise `FalseValue`
- `set_as_attr`: if `expression` yields non-empty (non-null) value, set the value as a client attribute.
- `{allow, all, subscribe, ["gb/${client_attrs.allowed_username}/#"]}.`: this rule only matches clients which have `allowed_username` attribute.

**Comment by zmstone** [comment]
To check `username` against certificate common name `cn` (not `cert_common_name` in this case),
I would recommend enterprise edition's client-info authentication: https://docs.emqx.com/en/emqx/latest/access-control/authn/cinfo.html#client-info-authentication -- this will reject bad clients from the very beginning.

Otherwise below expression can help to extract a client attribute and use in ACL rules.
```
iif(str_eq(username,cn), iif(regex_match(coalesce(username,''),'GBOX[0-9]{6}'),username,''), '')
```
